# Árbol de decisión

Según el siguiente árbol de decisión, analice y responda:

![Obama-Clinton decision tree](https://github.com/justmarkham/DAT8/raw/226791169b1cc6df8e8845c12e34e748d5ffaa85/notebooks/images/obama_clinton_tree.jpg)

## 1.1 ¿Qué son las observaciones? ¿Cuántas observaciones hay?


Las observaciones hacen referencia a la cantidad de condados en los que ganó cada candidato (Obama & Clinton).

La cantidad de observaciones corresponde al total de condados presentes en el estudio, que se distribuyen de la siguiente forma:
Total: 2241
Clinton: 1210
Obama: 1031

## 1.2 ¿Cuál es la variable de respuesta?


La cantidad de condados en los que ganaron cada uno de los candidatos.

## 1.3 ¿Cuáles son las características?¿Cuál es la característica más predictiva?


1. County more than 20% black.
2. High school graduation rate higher than 78%.
3. High school graduation rate higher than 87%.
4. County Ubication. 
5. Households poor in 2000.
6. Population density.
7. Bush beated Kerry in 2004.

- La característica 1. es la más predictiva.


## 1.4 ¿Cuál es la predicción de clase para el siguiente condado con los siguientes datos: 15% de afroamericanos, 90% de tasa de graduación en la escuela secundaria, ubicado en el Sur, alta pobreza, alta densidad de población?


La predicción de clase para los condados con esas caraterísticas es que Obama gana 185 a 36. 

## 1.5 ¿Cuál es el chance predicha para ese mismo condado anteriormente descrito?

Los chances predichos para ese condado para cada uno de los candidatos son:

In [95]:
Chanceo = 185/36 ; Chanceo

5.138888888888889

In [96]:
Chanceh = 36/185; Chanceh

0.1945945945945946

- Es 5,13 más problable que gane Obama  aque gane Clinton; análogamente es 0,20 veces más probables que gane Clinton a que lo haga Obama.
- Es más grande el chance que en estos condados gane Obama a que gane Clinton.

## SVM


Para esta tarea consideramos un conjunto de observaciones sobre una serie de variedades de vino tinto y blanco que implican sus propiedades químicas y su clasificación por parte de los catadores. La industria del vino ha experimentado un gran crecimiento en los últimos tiempos debido al aumento del consumo social. El precio del vino depende de un concepto bastante abstracto de apreciación del vino por parte de los catadores, cuya opinión puede tener un alto grado de variabilidad. El precio del vino depende en cierta medida de este factor tan volátil. Otro factor clave en la certificación y evaluación de la calidad del vino son las pruebas fisicoquímicas, que se realizan en laboratorio y tienen en cuenta factores como la acidez, el nivel de pH, la presencia de azúcar y otras propiedades químicas. Para el mercado del vino, sería interesante que la calidad humana de la cata pudiera relacionarse con las propiedades químicas del vino para que el proceso de certificación y evaluación de la calidad estuviera más controlado.

Se dispone de dos conjuntos de datos, uno de los cuales se refiere al vino tinto y cuenta con 1.599 variedades diferentes, y el otro, al vino blanco, con 4.898 variedades. Todos los vinos se producen en una zona concreta de Portugal. Se recogen datos sobre 12 propiedades diferentes de los vinos, una de las cuales es la calidad, basada en datos sensoriales, y el resto son propiedades químicas de los vinos, como la densidad, la acidez, el contenido de alcohol, etc. Todas las propiedades químicas de los vinos son variables continuas. La calidad es una variable ordinal con una clasificación posible de 1 (peor) a 10 (mejor). Cada variedad de vino es catada por tres catadores independientes y la clasificación final asignada es la mediana de la clasificación dada por los catadores.

Se espera que un modelo predictivo desarrollado a partir de estos datos sirva de orientación a los viñedos en cuanto a la calidad y el precio que se espera obtener de sus productos sin depender en gran medida de la volatilidad de los catadores.


In [97]:
import pandas as pd
import numpy as np

In [98]:
data_r = pd.read_csv('https://raw.githubusercontent.com/prof-tatiana-gutierrez/DatosMD/master/Wine_data_red.csv')
data_w = pd.read_csv('https://raw.githubusercontent.com/prof-tatiana-gutierrez/DatosMD/master/Wine_data_white.csv')

In [99]:
data = data_w.assign(type = 'white')

data = data.append(data_r.assign(type = 'red'), ignore_index=True)
data.sample(5)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
300             6.1              0.41         0.00             1.6      0.063   
1140            8.5              0.16         0.33             1.0      0.076   
2474            6.1              0.27         0.31             1.5      0.035   
5902            8.2              0.43         0.29             1.6      0.081   
4352            7.3              0.28         0.35             1.6      0.054   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
300                  36.0                  87.0  0.99140  3.27       0.67   
1140                 17.0                  57.0  0.99210  3.14       0.46   
2474                 17.0                  83.0  0.99076  3.32       0.44   
5902                 27.0                  45.0  0.99603  3.25       0.54   
4352                 31.0                 148.0  0.99178  3.18       0.47   

      alcohol  quality   type  
300      10.8        6  white  
1140     10.6        6  white  
2474     11.1        7  white  
5902     10.3        5    red  
4352     10.7        5  white

In [100]:
data.quality.value_counts()

6    2836
5    2138
7    1079
4     216
8     193
3      30
9       5
Name: quality, dtype: int64

In [101]:
data.type.value_counts()

white    4898
red      1599
Name: type, dtype: int64

# Ejercicio 2.1

Mostrar la tabla de frecuencias de la calidad por tipo de vino.

In [102]:
  pd.crosstab (index = data['quality'], columns = data['type'])

type     red  white
quality            
3         10     20
4         53    163
5        681   1457
6        638   2198
7        199    880
8         18    175
9          0      5

# Ejercicio 2.2

Teniendo en cuenta todas las variables:

* Crear un objetivo binario para cada tipo de vino, haciendo la transformación necesaria para ingresar la información al modelo.
* Crear una SVM que prediga el tipo de vino
* Evalúe con F1, AUC-ROC y log-loss


In [103]:
# Creación d ela indicadora
data['Wine'] = data['type'].apply(lambda x:1
                                   if x == 'red'  else 2)
#data.head(5)
#data.Wine.value_counts()

In [104]:
# Predicción del tipo de vino mediande SVM.
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split
x = data.drop(['type','Wine'],axis =1)
y = data['Wine']
X_train, X_test, y_train, y_test = train_test_split(x,y,test_size=0.2, random_state=111211)
# Modelo
modSVC = SVC(kernel='linear')
modSVC.fit(X_train, y_train)

SVC(kernel='linear')

In [105]:
# F1, AUC-ROC y log-loss
from sklearn import metrics
pred1 = modSVC.predict(X_test)

f1=np.sqrt(metrics.f1_score(y_test,pred1))
auc=metrics.roc_auc_score(y_test,pred1)
ll=np.sqrt(metrics.log_loss(y_test,pred1))

klin=pd.DataFrame([f1,auc,ll], index=['F1','AUC','Log-Loss']).transpose()
klin

F1       AUC  Log-Loss
0  0.988554  0.982357  2.969996

# Ejercicio 2.3

* Probar la SVM utilizando los diferentes kernels ('poly', 'rbf', 'sigmoid')
* Evalúe con F1, AUC-ROC y log-loss


In [106]:
# KERNEL 'POLY'
from sklearn import metrics

# Modelo.
modSVCp = SVC(kernel='poly')
modSVCp.fit(X_train, y_train)

# Predicciones.
pred2 = modSVCp.predict(X_test)

# Métricas.
f11=np.sqrt(metrics.f1_score(y_test,pred2))
auc1=metrics.roc_auc_score(y_test,pred2)
ll1=np.sqrt(metrics.log_loss(y_test,pred2))

kp = pd.DataFrame([f11,auc1,ll1], index=['F1','AUC','Log-Loss']).transpose()
kp

F1       AUC  Log-Loss
0  0.91993  0.894398  2.969996

In [107]:
# KERNEL 'RBF'
from sklearn import metrics

# Modelo.
modSVCr = SVC(kernel='rbf')
modSVCr.fit(X_train, y_train)

# Predicciones.
pred3 = modSVCr.predict(X_test)

# Métricas.
f12=np.sqrt(metrics.f1_score(y_test,pred3))
auc2=metrics.roc_auc_score(y_test,pred3)
ll2=np.sqrt(metrics.log_loss(y_test,pred3))

kr = pd.DataFrame([f12,auc2,ll2], index=['F1','AUC','Log-Loss']).transpose()
kr

F1      AUC  Log-Loss
0  0.932808  0.89535  2.969996

In [108]:
# KERNEL 'SIGMOID'
from sklearn import metrics

# Modelo.
modSVCs = SVC(kernel='sigmoid')
modSVCs.fit(X_train, y_train)

# Predicciones.
pred4 = modSVCs.predict(X_test)

# Métricas.
f13=np.sqrt(metrics.f1_score(y_test,pred4))
auc3=metrics.roc_auc_score(y_test,pred4)
ll3=np.sqrt(metrics.log_loss(y_test,pred4))

ks = pd.DataFrame([f13,auc3,ll3], index=['F1','AUC','Log-Loss']).transpose()
ks

F1       AUC  Log-Loss
0  0.700161  0.657672  2.969996

Al observar las métricas que se tienen como resultado de la estimación de los diferentes modelos al variar el kernel; se tiene que el mejor de ellos es el SVM con Kernel = 'rbf' (con f1:0.932808, AUC-ROC:0.89535 ,Log-Loss:		2.969996).

# Ejercicio 2.4
* Utilizando el mejor SVM del punto anterior encontrar los parámetros que da el mejor rendimiento teniendo en cuenta los siguientes hiperparámetros
'C': [0.1, 1, 10, 100, 1000], 'gamma': [0.01, 0.001, 0.0001]
* Evalúe cada modelo con F1, AUC-ROC y log-loss

Nota: Para los valores de C y gamma, deben realizar todas las posibles combinaciones entre estas dos variables.

In [109]:
# Parámetros a iterar.
Cp = [0.1, 1, 10, 100, 1000]
G =  [0.01, 0.001, 0.0001]

# Modelo con Kernel Rbf
F1 = [];AUC=[];LL=[];Cu=[];Gu=[]
for i in range(len(Cp)):
  for j in range(len(G)):
    Model = SVC(kernel = 'rbf', C = Cp[i], gamma = G[j])
    Model.fit(X_train, y_train)
    Prediction = Model.predict(X_test)
    F1.append(np.sqrt(metrics.f1_score(y_test,Prediction)))
    AUC.append(metrics.roc_auc_score(y_test, Prediction))
    LL.append(np.sqrt(metrics.log_loss(y_test,Prediction)))
    Cu.append(Cp[i])
    Gu.append(G[j])

In [110]:
# Métricas
M = pd.DataFrame([F1,AUC,LL,Cu,Gu], index = ['F1','AUC','Log-Loss','C','G']).transpose()
M.sort_values(['F1','AUC'],ascending=False)

F1       AUC  Log-Loss       C       G
13  0.990923  0.987865  2.969996  1000.0  0.0010
14  0.987804  0.981841  2.969996  1000.0  0.0001
10  0.984685  0.976806  2.969996   100.0  0.0010
12  0.979363  0.971211  2.969996  1000.0  0.0100
9   0.978689  0.971684  2.969996   100.0  0.0100
6   0.967695  0.949137  2.969996    10.0  0.0100
11  0.965124  0.944619  2.969996   100.0  0.0001
7   0.954815  0.928526  2.969996    10.0  0.0010
8   0.945062  0.913939  2.969996    10.0  0.0001
3   0.943186  0.910927  2.969996     1.0  0.0100
4   0.937463  0.901891  2.969996     1.0  0.0010
5   0.935523  0.898879  2.969996     1.0  0.0001
1   0.930544  0.893800  2.969996     0.1  0.0010
0   0.929601  0.889842  2.969996     0.1  0.0100
2   0.922795  0.882742  2.969996     0.1  0.0001

# Ejercicio 2.5

Compare los resultados con todos los anteriores modelos y eliga ¿Cuál algoritmo con cuáles hiperparámetros es el que tiene mejor rendimiento según la puntuación F1, AUC-ROC y log-loss?

Según los resultados antriores; el mejor modelo es un SVM con las siguientes características:
- C: 100
- Gamma: 0.0010
- Kernel: 'rbf'
Con el que se obtienen las siguientes métricas:
- F1: 0.990923
- AUC-ROC: 0.987865
- Log-Loss: 2.969996 * Profe, no tengo idea de porqué me da el mismo Log-Loss en todo* :(